src/depackers/arcfs.c

```diff
@@ -1,212 +1,335 @@
+   *   The timestamp counts 10ms increments from epoch 1900-01-01.
+   *   This is supposed to be stored when the top 12 bits of load_offset are 0xFFF.
+   * - Likewise, when the top 12 bits of load_offset are 0xFF, bits 8 through 19
+   *   in load_offset are supposed to be the RISC OS filetype.
+   */
+
+  /*  0 */ arc_uint8 method;
+  /*  1 */ char filename[12];
+  /* 12 */ arc_uint32 uncompressed_size;
+  /* 16    arc_uint32 load_offset; */ /* Low byte -> high byte of the 40-bit timestamp. */
+  /* 20    arc_uint32 exec_offset; */ /* Low portion of the 40-bit timestamp. */
+  /* 24    arc_uint32 attributes; */
+  /* 28 */ arc_uint32 compressed_size;
+  /* 32    arc_uint32 info; */
+  /* 36 */
+
+  /* Unpacked fields */
+  arc_uint16 crc16;
+  arc_uint8  compression_bits;
+  arc_uint8  is_directory;
+  arc_uint32 value_offset;
 };
 
+static int arcfs_check_magic(const unsigned char *buf)
+{
+  return memcmp(buf, "Archive\x00", 8) ? -1 : 0;
+}
 
-static int read_file_header(HIO_HANDLE *in, struct archived_file_header_tag *hdrp)
+static int arcfs_read_header(struct arcfs_data *data, HIO_HANDLE *f)
 {
-	int hlen, start /*, ver*/;
-	int i;
-
-	if (hio_seek(in, 8, SEEK_CUR) < 0)		/* skip magic */
-		return -1;
-	hlen = hio_read32l(in) / 36;
-	if (hio_error(in) != 0) return -1;
-	if (hlen < 1) return -1;
-	start = hio_read32l(in);
-	if (hio_error(in) != 0) return -1;
-	/*ver =*/ hio_read32l(in);
-	if (hio_error(in) != 0) return -1;
-
-	hio_read32l(in);
-	if (hio_error(in) != 0) return -1;
-	/*ver =*/ hio_read32l(in);
-	if (hio_error(in) != 0) return -1;
-
-	if (hio_seek(in, 68, SEEK_CUR) < 0)	/* reserved */
-		return -1;
-
-	for (i = 0; i < hlen; i++) {
-		int x = hio_read8(in);
-		if (hio_error(in) != 0) return -1;
-
-		if (x == 0)			/* end? */
-			break;
-
-		hdrp->method = x & 0x7f;
-		if (hio_read(hdrp->name, 1, 11, in) != 11) {
-			return -1;
-		}
-		hdrp->name[12] = 0;
-		hdrp->orig_size = hio_read32l(in);
-		if (hio_error(in) != 0) return -1;
-		hio_read32l(in);
-		if (hio_error(in) != 0) return -1;
-		hio_read32l(in);
-		if (hio_error(in) != 0) return -1;
-		x = hio_read32l(in);
-		if (hio_error(in) != 0) return -1;
-		hdrp->compressed_size = hio_read32l(in);
-		if (hio_error(in) != 0) return -1;
-		hdrp->offset = hio_read32l(in);
-		if (hio_error(in) != 0) return -1;
-
-		if (x == 1)			/* deleted */
-			continue;
-
-		if (hdrp->offset & 0x80000000)		/* directory */
-			continue;
-
-		hdrp->crc = x >> 16;
-		hdrp->bits = (x & 0xff00) >> 8;
-		hdrp->offset &= 0x7fffffff;
-		hdrp->offset += start;
-
-		/* Max allowed compression bits value is 16 for method FFh. */
-		if (hdrp->method > 2 && hdrp->bits > 16)
-			return -1;
-
-		return 0;
-	}
-
-	/* no usable files */
-	return -1;
+  arc_uint8 buffer[ARCFS_HEADER_SIZE];
+
+  if(hio_read(buffer, 1, ARCFS_HEADER_SIZE, f) < ARCFS_HEADER_SIZE)
+  {
+#ifdef ARCFS_DEBUG
+    debug("short read in header\n");
+#endif
+    return -1;
+  }
+
+  if(arcfs_check_magic(buffer) < 0)
+  {
+#ifdef ARCFS_DEBUG
```

Comment:
The ifdefs need indenting for consistency with other arc sources

Reply:
OK, fixed.